src/state.rs

```diff
@@ -370,80 +371,109 @@ impl TabExpandedString {
+/// meaning that is the result of using an exponentially weighted estimator (as described above) to
+/// estimate the value of another exponentially weighted estimator, which estimates the value of
+/// the raw data.
+///
+/// The purpose of this extra smoothing step is to reduce instantaneous fluctations in the estimate
+/// when large updates are received. Without this, estimates might have a large spike followed by a
+/// slow asymptotic approach to zero (until the next spike).
+#[derive(Debug)]
 pub(crate) struct Estimator {
-    steps: [f64; 16],
-    pos: u8,
-    full: bool,
+    smoothed_steps_per_sec: f64,
+    double_smoothed_steps_per_sec: f64,
     prev: (u64, Instant),
+    start: Instant,
 }
 
 impl Estimator {
     fn new(now: Instant) -> Self {
         Self {
-            steps: [0.0; 16],
-            pos: 0,
-            full: false,
+            smoothed_steps_per_sec: 0.0,
+            double_smoothed_steps_per_sec: 0.0,
             prev: (0, now),
+            start: now,
         }
     }
 
     fn record(&mut self, new: u64, now: Instant) {
-        let delta = new.saturating_sub(self.prev.0);
-        if delta == 0 || now < self.prev.1 {
+        // sanity check: don't record data if time or steps have not advanced
+        if new <= self.prev.0 || now < self.prev.1 {
             // Reset on backwards seek to prevent breakage from seeking to the end for length determination
             // See https://github.com/console-rs/indicatif/issues/480
             if new < self.prev.0 {
+                self.prev.0 = new;
                 self.reset(now);
             }
             return;
         }
 
-        let elapsed = now - self.prev.1;
-        let divisor = delta as f64;
-        let mut batch = 0.0;
-        if divisor != 0.0 {
-            batch = duration_to_secs(elapsed) / divisor;
-        };
+        let delta_steps = new - self.prev.0;
+        let delta_t = duration_to_secs(now - self.prev.1);
 
-        self.steps[self.pos as usize] = batch;
-        self.pos = (self.pos + 1) % 16;
-        if !self.full && self.pos == 0 {
-            self.full = true;
-        }
+        // the rate of steps we saw in this update
+        let new_steps_per_second = delta_steps as f64 / delta_t;
 
-        self.prev = (new, now);
-    }
+        // update the estimate: a weighted average of the old estimate and new data
+        let weight = estimator_weight(delta_t);
+        self.smoothed_steps_per_sec =
+            self.smoothed_steps_per_sec * weight + new_steps_per_second * (1.0 - weight);
 
-    pub(crate) fn reset(&mut self, now: Instant) {
-        self.pos = 0;
-        self.full = false;
-        self.prev = (0, now);
-    }
+        // Get an unbiased estimate of `smoothed_steps_per_sec` to serve as the data source for the
+        // double smoothed estimate. See comment on debiasing in `steps_per_second` for details.
```

Comment:
Even after consuming the `steps_per_second` documentation, I feel like "unbiased" and "debias" are bit too jargony/abstract/vague for me. Could you restate the relevant commits in more concrete/simpler terms?

Reply:
Sure; I'm doing my best to make it as comprehensive as possible without overloading the comments with paragraphs of information. Tell me what you'd like to see in the comments, and I'll do my best to make some changes.

Here's what's happening:

An exponentially weighted estimate is a single value that is iteratively updated over time. At each update, the previous value of the estimate is re-weighted according to its age, with new data receiving the rest of the weight. The raw value of the estimate therefore always reflects the *assumption* that all previous data in the estimate has been weighted according to its age. This is a false assumption.

When the first bit of data comes in, e.g. at t=1, then because the update process is iterative, the previous value of the estimator (which is arbitrary) will receive a weight appropriate for data that has aged by one second. With the default exponent weight of 15s, this means that the arbitrary initial estimator value will have a weight of ~86% in the stored estimate value. This is wrong - it should have weight zero in the estimate - because whatever value the estimator has before t=0 is arbitrary.

Another way of expressing this is to say that a naive iteratively updated exponentially weighted estimator is *biased* towards its initial value.

Fortunately, there is a very easy way to correct for this bias, and return a corrected (debiased) estimate. Included in the naive (biased) estimate is an initial value `x` with a total weight of `estimator_weight(delta_t_start)`. Thinking creatively, if we set the initial value of the estimate to 0, then removing the weight that this value has in the estimate is as trivial as dividing the estimate by the weight of all the data *since* `t=0`! So that's what we do. Whenever we need an estimate that treats the initial estimator value as having zero weight, we *debias* the raw estimator value.

-------------------

Do you think that sufficiently explains the idea? I could include the entire text in a comment if you think it's worth it.

-------------------

Here's a graph of the weight that the initial value has for an estimate of age `t`:

![initial_value_weight](https://github.com/console-rs/indicatif/assets/3870410/f2f80395-7b06-4105-b998-c2a164d199fa)

As you can see, there's no point at which this weight equals zero, but debiasing is most useful in the first 10-15 seconds. Otherwise you start with a value that is potentially very wrong, and slowly correct it. (Or, following a suggestion you made in a previous comment, you set the initial value of the estimator to be equal to the value of the first piece of data that comes in. But this first piece of data is usually *very* biased in my experience with file transfers, and getting a true estimate by dividing out the initial weight is cheap enough that it makes sense to just do that.)

Reply:
I think adding a lot of what you've written here as comments is exactly right. I still feel like the use of "debias" is jargony and maybe unnecessary if we can use something less specific like "corrected" instead, but I'm also fine with keeping debias if you think it's better for the text you're writing here.

Looking forward to merging this soon!